billing/loader.go

```diff
@@ -0,0 +1,236 @@
+	InvoicePeriod time.Time
+	ReportType    int
+	Fields        []string
+	csvReader     *csv.Reader
+	Mapper        fieldMapper
+}
+
+type billingRecord struct {
+	Id               string
+	ReportName       string
+	InvoiceID        string
+	PayerAccountId   uint64
+	LinkedAccountId  uint64
+	RecordType       string
+	RecordId         string
+	ProductName      string
+	RateId           int64
+	SubscriptionId   int64
+	PricingPlanId    int64
+	UsageType        string
+	Operation        string
+	AvailabilityZone string
+	ReservedInstance string
+	ItemDescription  string
+	UsageStartDate   string
+	UsageEndDate     string
+	UsageQuantity    float64
+	BlendedRate      float64
+	BlendedCost      float64
+	UnBlendedRate    float64
+	UnBlendedCost    float64
+	ResourceId       string
+	Tags             map[string]string
+}
+
+func (l *Loader) parseRecord(in chan []string, out chan *billingRecord, report *billingReport) {
+	var tagMatcher = regexp.MustCompile("(user|aws):.*")
+	var fieldTypes = map[string]func(s string, report *billingReport) interface{}{
+		"PayerAccountId":  l.parseUint,
+		"LinkedAccountId": l.parseUint,
+		"RecordInt":       l.parseUint,
+		"RateId":          l.parseInt,
+		"SubscriptionId":  l.parseInt,
+		"PricingPlanId":   l.parseInt,
+		"UsageStartDate":  l.parseDate,
+		"UsageEndDate":    l.parseDate,
+		"UsageQuantity":   l.parseFloat,
+		"BlendedRate":     l.parseFloat,
+		"BlendedCost":     l.parseFloat,
+		"UnBlendedRate":   l.parseFloat,
+		"UnBlendedCost":   l.parseFloat,
+	}
+
+	for values := range in {
+		var record = &billingRecord{}
+		record.Tags = make(map[string]string)
+		for i, field := range report.Fields {
+			if f, ok := fieldTypes[field]; ok {
+				l.assignField(record, field, f(values[i], report), false)
+			} else if tagMatcher.MatchString(field) {
+				value := map[string]string{l.parseTag(field, report).(string): values[i]}
+				l.assignField(record, field, value, true)
+			} else {
+				l.assignField(record, field, values[i], false)
+			}
+		}
+		out <- record
+	}
+	close(out)
+	l.wg.Done()
+}
+
+func (l *Loader) assignField(record *billingRecord, field string, value interface{}, tag bool) error {
+	if tag == false {
+		recordField := reflect.ValueOf(record).Elem().FieldByName(field)
+		if !recordField.IsValid() {
+			return fmt.Errorf("No such field '%s'", field)
+		}
+		if !recordField.CanSet() {
+			return fmt.Errorf("Field '%s' cannot be set.", field)
+		}
+		valueTyped := reflect.ValueOf(value)
+		if recordField.Type() != valueTyped.Type() {
+			return fmt.Errorf("Field '%s' didn't have type %s.", field, valueTyped.Type().Name())
+		}
+		recordField.Set(valueTyped)
+	} else {
+		for k, v := range value.(map[string]string) {
+			record.Tags[k] = v
+		}
+	}
+	return nil
+}
+
+func (l *Loader) openBillingReport(billingFile string) *billingReport {
+	var dateMonthMatcher = regexp.MustCompile(`\d{4}-\d{2}`)
+
+	invoicePeriod, err := time.Parse("2006-02", dateMonthMatcher.FindString(billingFile))
```

Comment:
The naming format is: '$ACCOUNT_ID-aws-billing-detailed-line-items-with-resources-and-tags-$YEAR-$MONTH.csv.zip' - but here it's the version without .zip extension.

I think I mentioned there: https://github.com/cycloidio/youdeploy/issues/17